src/Data/ElfEdit/Relocations/RISCV.hs

```diff
@@ -0,0 +1,444 @@
+
+pattern R_RISCV_SET8 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_SET8 = RISCV_RelocationType 54 -- S + A
+
+pattern R_RISCV_SET16 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_SET16 = RISCV_RelocationType 55 -- S + A
+
+pattern R_RISCV_SET32 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_SET32 = RISCV_RelocationType 56 -- S + A
+
+pattern R_RISCV_32_PCREL :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_32_PCREL = RISCV_RelocationType 57 -- S + A - P
+
+pattern R_RISCV_IRELATIVE :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_IRELATIVE = RISCV_RelocationType 58 -- `ifunc_resolver(B + A)`
+
+pattern R_RISCV_PLT32 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_PLT32 = RISCV_RelocationType 59 -- S + A - P
+
+pattern R_RISCV_SET_ULEB128 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_SET_ULEB128 = RISCV_RelocationType 60 -- S + A
+
+pattern R_RISCV_SUB_ULEB128 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_SUB_ULEB128 = RISCV_RelocationType 61 -- V - S - A
+
+pattern R_RISCV_TLSDESC_HI20 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_TLSDESC_HI20 = RISCV_RelocationType 62 -- S + A - P
+
+pattern R_RISCV_TLSDESC_LOAD_LO12 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_TLSDESC_LOAD_LO12 = RISCV_RelocationType 63 -- S - P
+
+pattern R_RISCV_TLSDESC_ADD_LO12 :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_TLSDESC_ADD_LO12 = RISCV_RelocationType 64 -- S - P
+
+pattern R_RISCV_TLSDESC_CALL :: (Eq (ElfWordType w), Num (ElfWordType w)) => RISCV_RelocationType w
+pattern R_RISCV_TLSDESC_CALL = RISCV_RelocationType 65
+
+riscvReloc :: RISCV_RelocationType w
+           -> String
+           -> Int
+           -> (RISCV_RelocationType w, (String,Int))
+riscvReloc tp nm c = (tp, (nm, c))
+
+-- These values are derived from Table 5 (Variables used in relocation fields) of
+-- https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/17038f12910bf6e0bc8bb12d3a2d09dce3f9152a/riscv-elf.adoc#relocations.
+
+none :: Int
+none = 0
+
+word6 :: Int
+word6 = 6
+
+word8 :: Int
+word8 = 8
+
+word16 :: Int
+word16 = 16
+
+word32 :: Int
+word32 = 32
+
+word64 :: Int
+word64 = 64
+
+-- The following variable types are described in Figure 2.3 (RISC-V base
+-- instruction formats showing immediate variants) and Table 12.1 (Compressed
+-- 16-bit RVC instruction formats) of
+-- https://riscv.org/wp-content/uploads/2017/05/riscv-spec-v2.2.pdf.
+--
+-- Note that the treatment of these variables is not correct. See
+-- https://github.com/GaloisInc/elf-edit/issues/39 for more information.
+
+bType :: Int
+bType = 32
```

Comment:
As noted here, the treatment of `bType` (and related definitions) is not correct. If you'd prefer, I could make these definitions `error` if they're used—I'm undecided which approach is better. (See https://github.com/GaloisInc/elf-edit/issues/39 for more details.)

Reply:
My inclination would be to use `error` on the basis that a clear indication of bad parsing is much better than silently providing bad data for downstream processing that will be very hard to trace back to this cause.